pkg/conference/participant/tracker.go

```diff
@@ -85,151 +90,98 @@ func (t *Tracker) RemoveParticipant(participantID ID) map[string]bool {
 // that has been published and that we must take into account from now on.
 func (t *Tracker) AddPublishedTrack(
 	participantID ID,
-	info webrtc_ext.TrackInfo,
-	simulcast webrtc_ext.SimulcastLayer,
-	metadata TrackMetadata,
-	outputTrack *webrtc.TrackLocalStaticRTP,
-) {
-	// If this is a new track, let's add it to the list of published and inform participants.
-	track, found := t.publishedTracks[info.TrackID]
-	if !found {
-		layers := []webrtc_ext.SimulcastLayer{}
-		if simulcast != webrtc_ext.SimulcastLayerNone {
-			layers = append(layers, simulcast)
-		}
+	remoteTrack *webrtc.TrackRemote,
+	metadata track.TrackMetadata,
+) error {
+	participant := t.participants[participantID]
+	if participant == nil {
+		return fmt.Errorf("participant %s does not exist", participantID)
+	}
 
-		t.publishedTracks[info.TrackID] = &PublishedTrack{
-			Owner:         participantID,
-			Info:          info,
-			Layers:        layers,
-			Metadata:      metadata,
-			OutputTrack:   outputTrack,
-			Subscriptions: make(map[ID]subscription.Subscription),
+	// If this is a new track, let's add it to the list of published and inform participants.
+	if published, found := t.publishedTracks[remoteTrack.ID()]; found {
+		if err := published.AddPublisher(remoteTrack); err != nil {
+			return err
 		}
 
-		return
+		return nil
 	}
 
-	// If it's just a new layer, let's add it to the list of layers of the existing published track.
-	fn := func(layer webrtc_ext.SimulcastLayer) bool { return layer == simulcast }
-	if simulcast != webrtc_ext.SimulcastLayerNone && slices.IndexFunc(track.Layers, fn) == -1 {
-		track.Layers = append(track.Layers, simulcast)
-		t.publishedTracks[info.TrackID] = track
+	published, err := track.NewPublishedTrack(
+		participantID,
+		participant.Peer.RequestKeyFrame,
+		remoteTrack,
+		metadata,
+		participant.Logger,
+	)
+	if err != nil {
+		return err
 	}
+
+	// Wait for the track to complete and inform the conference about it.
+	go func() {
+		// Wait for the track to complete.
+		<-published.Done()
+
+		// Inform the conference that the track is gone. Or stop the go-routine if the conference stopped.
+		select {
+		case t.publishedTrackStopped <- TrackStoppedMessage{remoteTrack.ID(), participantID}:
+		case <-t.conferenceEnded:
+		}
```

Comment:
Select chooses only one case at random if both are ready. The secound case is to to prevent this goroutine from blocking indefinitely? Then the channel `t.publishedTrackStopped` should never be closed, or we'll panic at some point. on the other hand an unclosed channel cannot be cleaned up by the garbage collector.

Reply:
If we were sure thats `t.conferenceEnded` is closed before `t.publishedTrackStopped` is closed and cleaned up, we could doing something like this:
```go
select {
   case <-t.conferenceEnded: 
      return
   default: 
      t.publishedTrackStopped <- TrackStoppedMessage{remoteTrack.ID(), participantID}:	
}
```

Reply:
A condition thats this not blocking endless is thats all messages from `t.publishedTrackStopped` are read before `close(t.publishedTrackStopped)`

Reply:
We often have that a send channel depends on another open state (here is conference still existsing)

What do you think about this? Has this side effects?

```go
trackIsDone = c.tracker.AddPublishedTrack(sender, msg.RemoteTrack, trackMetadata)
go func() {
   select {
      case <- trackIsDone: 
           c.tracker.RemovePublishedTrack(id track.TrackID)
      case <- t.conferenceEnded:
          // stop track
   }
}
```

Reply:
> Are you sure about that? - AFAIK, the channels are freed by the garbage collector once they are not used.

No. I was wrong with that. I was talking about buffered channels. But here as well, when the channel is not
referenced, the values in the channel buffer are also not referenced. An thy get cleaned from the gcc  (Ian Lance Taylor, Dave Cheney, I think we can trust them :D)

So we have not to close channels, from this point of few.